src/parser/expression.lisp

```diff
@@ -875,6 +945,225 @@
+                     (cst:atom (cst:second form))
+                     (keywordp (cst:raw (cst:second form))))
+                (cst:raw (cst:second form))))
+            (unparsed-body
+              (cond (label
+                     ;; (loop label)
+                     (if (cst:consp (cst:nthrest 2 form))
+                         (cst:nthrest 2 form)
+                         (error 'parse-error
+                                :err (coalton-error
+                                      :span (cst:source form)
+                                      :file file
+                                      :highlight :end
+                                      :message "Malformed loop expression"
+                                      :primary-note "expected a loop body"))))
+                    ;; (loop)
+                    ((not (cst:consp (cst:rest form)))
+                     (error 'parse-error
+                            :err (coalton-error
+                                  :span (cst:source form)
+                                  :file file
+                                  :highlight :end
+                                  :message "Malformed loop expression"
+                                  :primary-note "expected a loop body or label")))
+
+                    (t
+                     (cst:rest form))))
+            (*escapable-loop-context* t)
+            (*loop-label-context*
+              (if label 
+                  (cons label *loop-label-context*)
+                  *loop-label-context*)))
+
+       (make-node-loop
+        :source (cst:source form)
+        :label label
+        :body (parse-body unparsed-body form file))))
+
+    ((and (cst:atom (cst:first form))
+          (eq 'coalton:break (cst:raw (cst:first form))))
+     (let ((label
+             ;; (break)
+             (and (cst:consp (cst:rest form))
+                  (if (and (cst:atom (cst:second form))
+                           (keywordp (cst:raw (cst:second form)))) 
+                      (cst:raw (cst:second form))
+                      ;; (break not-a-keyword)
+                      (error 'parse-error
+                             :err (coalton-error
+                                   :span (cst:source form)
+                                   :file file
+                                   :highlight :end
+                                   :message "Invalid break expression"
+                                   :primary-note "expected a keyword label"))))))                 
+       (when label
+         (unless (member label *loop-label-context*)
+           (error 'parse-error
+                  :err (coalton-error
+                        :span (cst:source form)
+                        :file file
+                        :highlight :end
+                        :message "Invalid break expression"
+                        :primary-note (format nil "Label ~s not found in any enclosing loop" label)))))
```

Comment:
`"label not found in any enclosing loop"`

no need to pretty print the label name when it can be highlighted